src/context/component-hooks.js

```diff
@@ -0,0 +1,191 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {devAssert} from '../log';
+import {protectedNoInline} from './scheduler';
+
+const EMPTY_DEPS = [];
+
+/**
+ * @type {!./component.Component|undefined}
+ */
+let currentComponent;
+
+/**
+ * Sets the component as a current component and executes the callback. Clears
+ * out the current component after the callback is done.
+ *
+ * @param {!./component.Component} component
+ * @param {function()} callback
+ * @package
+ */
+export function withComponent(component, callback) {
+  currentComponent = component;
+  try {
+    callback();
+  } finally {
+    currentComponent = undefined;
+  }
+}
+
+/**
+ * @return {!./component.Component}
+ */
+function getComponent() {
+  return devAssert(currentComponent);
+}
+
+/**
+ * A reusable reference value hook. Mostly the same as the React's `useHook`.
+ *
+ * @param {T} def
+ * @return {{current: T}}
+ * @template T
+ */
+export function useRef(def = undefined) {
+  return getComponent().allocRef(def);
+}
+
+/**
+ * A hook to compute a persistent value. Mostly the same as the React's
+ * `useMemo` API.
+ *
+ * @param {function():T} compute
+ * @param {!Array} deps
+ * @return {T}
+ * @template T
+ */
+export function useMemo(compute, deps) {
+  const ref = useRef();
+  useSyncEffect(() => {
+    ref.current = compute();
+  }, deps);
+  return ref.current;
+}
+
+/**
+ * Functions similarly to `useMemo`, but besides being able to return a computed
+ * value, it also provides a cleanup callback.
+ *
+ * @param {function(!Node, ...?):({value: T, dispose: function()}|function())} factory
+ * @param {!Array=} deps
+ * @return {T}
+ * @template T
+ */
+export function useDisposable(factory, deps = undefined) {
```

Comment:
It's exactly where this comes from. Just simpler. Hooks are very composable and I'd like to see more of them. But I felt like `useService` is a bit too ambiguous: it sounds like more like "find me a service" rather than "memoize this factory". An alternative name, maybe, would be `useDisposableMemo()`?

Reply:
I used `useDisposableMemo` for now.